### DISCUSSION TITLE
'binary'-style params: should E3 be able to change state in params menu?

### DISCUSSION THREAD

**Discussion by dndrks** [monome/norns#1752] [General]
_brought up in the norns study Discord_

currently, `binary`-style parameters are able to be changed by both encoders and keys.
the visual UI is tuned to show state changes with keypresses -- eg. 'momentary' is a binary of whether a key is held or released. this interaction is difficult to model with E3, though, but this type is currently able to be manipulated via encoder turns. however, encoders do not do the necessary [UI toggling](https://github.com/monome/norns/blob/f8e9ec5c0d5f8e4434fe338c17814e037487917c/lua/core/menu/params.lua#L195-L199) for this param type.

it seems like there are two options:

- bake `binary`-type UI toggling into params menu E3 movement as well as keypresses, maybe even putting this into the `:delta` function for this param type
- restrict E3's affect to exclude `binary`-type params

@tehn , any preference?

**Comment by duncangeere** [comment]
Hey folks. I'm the one who brought this up. 

FWIW my expectation was that turning the encoder clockwise would toggle it on and turning it anticlockwise would toggle it off. That seems to be what happens behind the scenes, I think? But it doesn't visually reflect that in the UI. It would be nice if it did!

**Comment by dndrks** [comment]
thanks so much for the additional notes, @duncangeere!
just in case it helps anyone reading, `binary` parameter types have three behaviors: toggle, momentary, and trigger.

current K3 interactions with these behaviors:
- toggle: each press flips the state, release is ignored
- momentary: press enables, release disables
- trigger: press bangs the action, and a [timer disables the visual](https://github.com/monome/norns/blob/f8e9ec5c0d5f8e4434fe338c17814e037487917c/lua/core/menu/params.lua#L664-L669)

@duncangeere , your expectation makes a ton of sense for the 'toggle' behavior, but it feels like it wouldn't map meaningfully to the 'trigger' and 'momentary' behaviors.

do we maybe want E3 encoder turns to _only_ affect 'toggle'-style `binary` params? or do we want to restrict encoder behavior completely?

**Comment by duncangeere** [comment]
Yes! I should have been clearer that I was speaking entirely from a "I want to toggle things" pov. Thanks for mentioning that.

From a user perspective, E3 is how you interact with most? every? kind of parameter *except* binary ones, which is how I ended up discovering this bug in the first place - it didn't even occur to me to use K3 when I was in the params menu.

So my feeling is that there should be *some* kind of encoder support in all of them. How about:

- toggle: turning clockwise enables, turning anticlockwise disables
- momentary: any kind of turning enables and stopping the motion disables
- trigger: any kind of turning bangs the action, and a timer disables the visual. Maybe all triggers within the timer period after the first one are ignored?

**Comment by tehn** [comment]
toggle should be equivalent to 0/1 so this makes perfect sense regarding E3
behavior

i'm ambivalent about MOM/TRIG with encoders because it makes so much more
sense with a key: to the point where i think encoder actions should perhap
just be ignored on those?
